helix-core/src/test/java/org/apache/helix/task/TestDynamicTaskLoading.java

```diff
@@ -0,0 +1,251 @@
+    String workflowName = TestHelper.getTestMethodName();
+    submitWorkflow(workflowName, driver);
+
+    try {
+      // Wait for the workflow to either complete or fail.
+      TaskState finalState =
+          driver.pollForWorkflowState(workflowName, TaskState.COMPLETED, TaskState.FAILED);
+      AssertJUnit.assertEquals(finalState, TaskState.FAILED);
+    } catch (HelixException e) {
+      AssertJUnit.fail(e.getMessage());
+    }
+  }
+
+  @Test
+  public void testDynamicTaskLoadingNonexistingTaskClass() throws Exception {
+    // Add task definition information as a DynamicTaskConfig.
+    List<String> taskClasses = new ArrayList<String>();
+    taskClasses.add("com.mycompany.mocktask.RandomTask");
+    DynamicTaskConfig taskConfig =
+        new DynamicTaskConfig("Reindex", "src/test/resources/Reindex.jar", "1.0.0", taskClasses,
+            "com.mycompany.mocktask.MockTaskFactory");
+    String path = TaskConstants.DYNAMICALLY_LOADED_TASK_PATH + "/Reindex";
+    removePathIfExists(path);
+    _manager.getHelixDataAccessor().getBaseDataAccessor()
+        .create(path, taskConfig.getTaskConfigZNRecord(), AccessOption.PERSISTENT);
+
+    // Submit workflow
+    TaskDriver driver = new TaskDriver(_manager);
+    String workflowName = TestHelper.getTestMethodName();
+    submitWorkflow(workflowName, driver);
+
+    try {
+      // Wait for the workflow to either complete or fail.
+      TaskState finalState =
+          driver.pollForWorkflowState(workflowName, TaskState.COMPLETED, TaskState.FAILED);
+      AssertJUnit.assertEquals(finalState, TaskState.FAILED);
+    } catch (HelixException e) {
+      AssertJUnit.fail(e.getMessage());
+    }
+  }
+
+  @Test
+  public void testDynamicTaskLoadingNonexistingTaskFactory() throws Exception {
+    // Add task definition information as a DynamicTaskConfig.
+    List<String> taskClasses = new ArrayList<String>();
+    taskClasses.add("com.mycompany.mocktask.MockTask");
+    DynamicTaskConfig taskConfig =
+        new DynamicTaskConfig("Reindex", "src/test/resources/Reindex.jar", "1.0.0", taskClasses,
+            "com.mycompany.mocktask.RandomTaskFactory");
+    String path = TaskConstants.DYNAMICALLY_LOADED_TASK_PATH + "/Reindex";
+    removePathIfExists(path);
+    _manager.getHelixDataAccessor().getBaseDataAccessor()
+        .create(path, taskConfig.getTaskConfigZNRecord(), AccessOption.PERSISTENT);
+
+    // Submit workflow
+    TaskDriver driver = new TaskDriver(_manager);
+    String workflowName = TestHelper.getTestMethodName();
+    submitWorkflow(workflowName, driver);
+
+    try {
+      // Wait for the workflow to either complete or fail.
+      TaskState finalState =
+          driver.pollForWorkflowState(workflowName, TaskState.COMPLETED, TaskState.FAILED);
+      AssertJUnit.assertEquals(finalState, TaskState.FAILED);
+    } catch (HelixException e) {
+      AssertJUnit.fail(e.getMessage());
+    }
+  }
+
+  @AfterMethod
+  public void afterMethod() {
```

Comment:
Is it possible to use AfterClass instead to avoid redundant work of starting and shutting down threads?

Reply:
Regarding this and BeforeMethod vs. BeforeClass, I wanted to make sure no dynamically loaded class remains loaded between the tests. I tried with BeforeClass and AfterClass, and all tests would result in TaskState.COMPLETED whereas only one should complete and the others should fail.

Reply:
That's fair. Could you make sure you document the reason in your before method/after method ?